### PR TITLE
Import original celeryutils code from edx-platform.

### PR DESCRIPTION
**Pull request by jcdyer**
# [TNL-6454](https://openedx.atlassian.net/browse/TNL-6454) 

Subtask of [TNL-6294](https://openedx.atlassian.net/browse/TNL-6294).

* Make configuration work in freestanding context.
* Convert tests to idiomatic py.test tests
* Clean up docs and formatting

## Reviewers

- [ ] @nasthagiri 
- [ ] @efischer19

### CODE REVIEW

File: celeryutils/persist_on_failure.py [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+"""
+
+from __future__ import absolute_import, division, print_function, unicode_literals
+
+from celery import Task
+
+from .models import FailedTask
+
+
+# pylint: disable=abstract-method
```
Comment (efischer19) [inline celeryutils/persist_on_failure.py]: disable-abstract-**method** seems like an odd pylint warning to have above a class. I'm sure it's correct, but it's still weird.

File: test_utils/tasks.py [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+    if message:
+        raise ValueError(message)
+
+
+@app.task(base=persist_on_failure.PersistOnFailureTask)
+def passing_task():
+    """
+    This task always passes
+    """
+    return 5
```
Comment (efischer19) [inline test_utils/tasks.py]: Shouldn't the body of this task method look like this?
```
def passing_task():
     """
     This task always passes
     """
     pass
```


image

File: celeryutils/management/commands/cleanup_resolved_tasks.py [hunk 3]
```diff
@@ -0,0 +1,60 @@
+"""
+Reset persistent grades for learners.
```
Comment (jcdyer) [inline celeryutils/management/commands/cleanup_resolved_tasks.py]: Good point.

File: test_utils/tasks.py (hunk 2 above)
Comment (jcdyer) [inline test_utils/tasks.py]: No.

**Comment by jcdyer** [comment]
@nasthagiri @efischer19 I upgraded travis to test against both celery 3 and 4 (and only 4 against later django versions because I don't want to exponentially blow out test configurations, and I'm optimistic that we won't be on celery 3 for that long).   Latest version is RFR.

File: Makefile [hunk 4]
```diff
@@ -44,6 +44,11 @@ upgrade: ## update the requirements/*.txt files with the latest packages satisfy
 	pip-compile --upgrade -o requirements/quality.txt requirements/quality.in
 	pip-compile --upgrade -o requirements/test.txt requirements/base.in requirements/test.in
 	pip-compile --upgrade -o requirements/travis.txt requirements/travis.in
+	# Workaround for packages that break pip-tools due to python version specifiers somewhere upstream
+	# See https://github.com/nvie/pip-tools/issues/206
+	cat requirements/dev-extra.txt >> requirements/dev.txt
```
Comment (nasthagiri) [inline Makefile]: I don't remember running into this when I used the cookie-cutter for a modulestore repo.
Is there a JIRA ticket for addressing this?

File: celeryutils/persist_on_failure.py [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+class PersistOnFailureTask(Task):
+    """
+    Custom Celery Task base class that persists task data on failure.
+    """
+
+    abstract = True
+    typing = False
```
Comment (nasthagiri) [inline celeryutils/persist_on_failure.py]: Why is type-checking set to `False`?

File: celeryutils/persist_on_failure.py [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+from .models import FailedTask
+
+
+# pylint: disable=abstract-method
+class PersistOnFailureTask(Task):
+    """
+    Custom Celery Task base class that persists task data on failure.
+    """
+
+    abstract = True
```
Comment (nasthagiri) [inline celeryutils/persist_on_failure.py]: Why was it necessary to add `abstract` here?

File: celeryutils/persist_on_failure.py (hunk 1 above)
Comment (nasthagiri) [inline celeryutils/persist_on_failure.py]: agreed.  What is the fix that it's asking for?

File: requirements/base.txt [hunk 7]
```diff
@@ -4,4 +4,12 @@
 #
 #    pip-compile --output-file requirements/base.txt requirements/base.in
 #
-Django==1.10.5
+amqp==2.1.4               # via kombu
+billiard==3.5.0.2         # via celery
+celery==4.0.2
+django-model-utils==2.6.1
```
Comment (nasthagiri) [inline requirements/base.txt]: Why is this a different version from edx-platform?

File: requirements/base.txt [hunk 8]
```diff
@@ -4,4 +4,12 @@
...
+Django==1.10.5            # via django-model-utils, jsonfield
```
Comment (nasthagiri) [inline requirements/base.txt]: Isn't this also a different version from edx-platform?

**Comment by jcdyer** [comment]
@efischer19 Click on "Show all checks" below, then select your favorite travis run.

File: requirements/doc.txt [hunk 9]
```diff
@@ -5,21 +5,27 @@
 #    pip-compile --output-file requirements/doc.txt requirements/base.in requirements/doc.in
 #
 alabaster==0.7.9          # via sphinx
+amqp==2.1.4               # via kombu
```
Comment (nasthagiri) [inline requirements/doc.txt]: Why are all these additional requirements needed for Doc?

File: requirements/quality-extra.txt [hunk 10]
```diff
@@ -0,0 +1,2 @@
+pylint
+edx-lint                  # edX pylint rules and plugins
```
Comment (nasthagiri) [inline requirements/quality-extra.txt]: What is the purpose for the quality-extra.txt file?
These are both also in quality.txt.

Same question about dev-extra.txt above.

File: setup.cfg [hunk 11]
```diff
@@ -1,10 +1,12 @@
 [isort]
 line_length = 120
-known_edx =
+known_third_party = ddt,mock
```
Comment (nasthagiri) [inline setup.cfg]: Why was this deleted?

File: test_utils/tasks.py [hunk 12]
```diff
@@ -0,0 +1,26 @@
+"""
+Tasks used in tests
```
Comment (nasthagiri) [inline test_utils/tasks.py]: Hmm.. aren't these specific to `PersistOnFailureTask`?  Why are they pulled out into a generic test-utils?

File: celeryutils/persist_on_failure.py (hunk 1 above)
Comment (jcdyer) [inline celeryutils/persist_on_failure.py]: There is an abstract method defined on `celery.Task` (`run()`), that isn't overridden here, because this is also an abstract task base class, but celery uses its own annotation for abstract classes, so pylint doesn't understand it.

File: celeryutils/persist_on_failure.py (hunk 6 above)
Comment (jcdyer) [inline celeryutils/persist_on_failure.py]: It's a celery flag to prevent this task from getting registered as a callable task.

File: celeryutils/persist_on_failure.py (hunk 5 above)
Comment (jcdyer) [inline celeryutils/persist_on_failure.py]: In Celery 3, celery will blindly send the arguments you pass to your command, whether or not they are the right arguments.  The task will then fail if it can't handle them.  Celery 4 checks the arguments before it sends the task, and raises a TypeError if they don't check out.  This opts out of that behavior on Celery 4 (and has no effect on Celery 3). 

Submitting incorrectly called tasks allows us to to catch and persist those tasks in our FailedTask table.  Without this, the TypeError would occur on the app server, not the task worker, so the task would never get persisted.

File: requirements/base.txt (hunk 8 above)
Comment (jcdyer) [inline requirements/base.txt]: See base.in as well as tox.ini and .travis.yml.  Tests are run against the latest versions of django 1.8, 1.9, and 1.10.

File: requirements/quality-extra.txt (hunk 10 above)
Comment (jcdyer) [inline requirements/quality-extra.txt]: quality.txt = pip-compile quality.in + quality-extra.txt.  

So those requirements are in quality.txt *because* they are in quality-extra.txt.  The reason for including them there instead of in quality.in is because pip-compile comes up against the pip-tools bug mentioned in the Makefile.

See the specification for `make upgrade` in Makefile, as well as the docs for the example workflow in [pip-tools](https://github.com/nvie/pip-tools#example-usage-for-pip-compile) for the relationship between dev.in, dev-extra.txt and dev.txt.

File: celeryutils/persist_on_failure.py (hunk 5 above)
Comment (nasthagiri) [inline celeryutils/persist_on_failure.py]: excellent.  So this for forward compatibility with Celery 4.

File: requirements/doc.txt (hunk 9 above)
Comment (jcdyer) [inline requirements/doc.txt]: I'm not sure.  Perhaps sphinx needs the be able to import the app to build its docs?  The relationship between base.in and doc.txt is defined by the cookiecutter, so @jmbowman may be better equipped to answer the question.

File: requirements/quality-extra.txt (hunk 10 above)
Comment (nasthagiri) [inline requirements/quality-extra.txt]: So if quality.txt is a generated file, shouldn't it be excluded from committing into the repo?

File: setup.cfg (hunk 11 above)
Comment (jcdyer) [inline setup.cfg]: I added it because it isort was complaining about them, but that's because I hadn't installed my requirements yet.  Once I did that, it stopped complaining.

isort checks your import sort order, and it considers something a third-party package if it lives in lib/python*/site-packages, so without the packages being installed, it couldn't identify them properly.

File: test_utils/tasks.py (hunk 12 above)
Comment (jcdyer) [inline test_utils/tasks.py]: They are specific to the tests, so they shouldn't be installed with celeryutils, which is why they don't live in celeryutils.  I didn't put them in `/tests`, because the project structure seems designed to use that for actual test code. 

Also, the file is also used for tasks that test the logging decorator, once that PR comes in.

Note the /tests directory doesn't have a `__init__.py` file, and the tox file explicitly adds one for the purposes of running pylint against the tests, but removes it again at the end, which makes it less suitable for code that might get imported from different places.  @jmbowman might be able to shed light on the rationale for that decision.

File: Makefile (hunk 4 above)
Comment (jcdyer) [inline Makefile]: I'm not sure.  Jeremy is aware of the pip-tools issue, and I think it's fairly high on his platter, but I'm not sure if it exists in jira or not.

File: celeryutils/persist_on_failure.py (hunk 5 above)
Comment (jcdyer) [inline celeryutils/persist_on_failure.py]: Yes.  There's a test that calls the task with the wrong args, and without this flag it fails under celery 4.

File: requirements/base.txt (hunk 7 above)
Comment (jcdyer) [inline requirements/base.txt]: The .txt requirement files are autogenerated (by pip-tools) from the .in files for local development to enable replicable builds.  When edx-platform installs this app, it will use setup.py (which has loosely versioned or unversioned dependencies).

tox is set up to multiple runs against multiple versions of key requirements, particularly python, django, and celery (which affects several of celery's dependencies as well).

django-model-utils is not currently tested with different versions.  The library is currently unpinned in base.in (because it doesn't require a particular version), and shouldn't need to be upgraded separately when edx-platform wants to update its requirement.

File: requirements/quality-extra.txt (hunk 10 above)
Comment (jcdyer) [inline requirements/quality-extra.txt]: I think the rationale for having it included is to have repeatable builds (it only gets updated when you explicitly run `pip-compile` or `make upgrade`).  But if the library is imported in an external project, I'm not sure that goal is really acheived, since it will install the version from setup.py, not from base.txt.

File: requirements/doc.txt (hunk 9 above)
Comment (jmbowman) [inline requirements/doc.txt]: If you use `sphinx-apidoc` to auto-generate API docs for the Python code (which projects generated from the cookiecutter do by default), it actually tries to import each file in order to retrieve docstrings and such.  This typically fails unless all the imported libraries are available, hence the virtualenv used to run Sphinx also needs all the library's core dependencies installed.

File: requirements/quality-extra.txt (hunk 10 above)
Comment (jmbowman) [inline requirements/quality-extra.txt]: quality.txt is generated, but the output is different depending on when you run it (based on currently available library versions).  The idea is to use the exact same versions consistently until we make a deliberate choice to upgrade them, so we need the output at a particular point in time checked in.

File: test_utils/tasks.py (hunk 12 above)
Comment (jmbowman) [inline test_utils/tasks.py]: pytest [recommends](http://doc.pytest.org/en/latest/goodpractices.html#choosing-a-test-layout-import-rules) omitting `__init__.py` from test directories for a few different reasons, but without it pylint won't analyze anything in the directory (at least not without separately running it on each contained module and package).
